flink-ml-iteration/src/main/java/org/apache/flink/iteration/operator/HeadOperator.java

```diff
@@ -376,29 +379,37 @@ public void endInput() throws Exception {
         InputChannel inputChannel =
                 getContainingTask().getEnvironment().getAllInputGates()[0].getChannel(0);
 
-        boolean endOfPartitionReceived = false;
-        long lastTriggerCheckpointId = 0;
-        while (!endOfPartitionReceived && status != HeadOperatorStatus.TERMINATED) {
-            mailboxExecutor.tryYield();
-            Thread.sleep(200);
-
-            List<AbstractEvent> events = parseInputChannelEvents(inputChannel);
-
-            for (AbstractEvent event : events) {
-                if (event instanceof CheckpointBarrier) {
-                    CheckpointBarrier barrier = (CheckpointBarrier) event;
-                    if (barrier.getId() > lastTriggerCheckpointId) {
-                        getContainingTask()
-                                .triggerCheckpointAsync(
-                                        new CheckpointMetaData(
-                                                barrier.getId(), barrier.getTimestamp()),
-                                        barrier.getCheckpointOptions());
-                        lastTriggerCheckpointId = barrier.getId();
-                    }
-
-                } else if (event instanceof EndOfPartitionEvent) {
-                    endOfPartitionReceived = true;
-                }
+        AtomicBoolean endOfPartitionReceived = new AtomicBoolean(false);
+        AtomicLong lastTriggerCheckpointId = new AtomicLong(0);
+
+        while (!endOfPartitionReceived.get() && status != HeadOperatorStatus.TERMINATED) {
+            Future<Void> future =
+                    mailboxExecutor.submit(
```

Comment:
It introduces a spin loop. We typically should avoid spin loop due to its waste of CPU cycles.